modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/MoveKBObjectMVCActionCommand.java

```diff
@@ -52,31 +64,114 @@ protected void doProcessAction(
 			ActionRequest actionRequest, ActionResponse actionResponse)
 		throws Exception {
 
-		long resourceClassNameId = ParamUtil.getLong(
-			actionRequest, "resourceClassNameId");
-		long resourcePrimKey = ParamUtil.getLong(
-			actionRequest, "resourcePrimKey");
-		long parentResourcePrimKey = ParamUtil.getLong(
-			actionRequest, "parentResourcePrimKey",
-			KBFolderConstants.DEFAULT_PARENT_FOLDER_ID);
-
-		long kbArticleClassNameId = _portal.getClassNameId(
-			KBArticleConstants.getClassName());
+		boolean dragAndDrop = ParamUtil.getBoolean(
+			actionRequest, "dragAndDrop");
 
-		if (resourceClassNameId == kbArticleClassNameId) {
+		try {
+			long resourceClassNameId = ParamUtil.getLong(
+				actionRequest, "resourceClassNameId");
+			long resourcePrimKey = ParamUtil.getLong(
+				actionRequest, "resourcePrimKey");
 			long parentResourceClassNameId = ParamUtil.getLong(
 				actionRequest, "parentResourceClassNameId",
 				_portal.getClassNameId(KBFolderConstants.getClassName()));
-			double priority = ParamUtil.getDouble(actionRequest, "priority");
+			long parentResourcePrimKey = ParamUtil.getLong(
+				actionRequest, "parentResourcePrimKey",
+				KBFolderConstants.DEFAULT_PARENT_FOLDER_ID);
+
+			long kbArticleClassNameId = _portal.getClassNameId(
+				KBArticleConstants.getClassName());
+
+			boolean movable = true;
+
+			if (resourceClassNameId == kbArticleClassNameId) {
+				double priority = ParamUtil.getDouble(
+					actionRequest, "priority");
+
+				if (_isDragAndDrop(dragAndDrop)) {
+					KBArticle kbArticle = _kbArticleService.getLatestKBArticle(
+						resourcePrimKey, WorkflowConstants.STATUS_ANY);
+
+					if (kbArticle.getParentResourcePrimKey() ==
+							parentResourcePrimKey) {
+
+						movable = false;
+					}
+
+					priority = kbArticle.getPriority();
+				}
+
+				if (movable) {
+					_kbArticleService.moveKBArticle(
+						resourcePrimKey, parentResourceClassNameId,
+						parentResourcePrimKey, priority);
+				}
+			}
+			else {
+				if (_isDragAndDrop(dragAndDrop)) {
+					if (parentResourceClassNameId == kbArticleClassNameId) {
+						_errorMessage(
+							actionRequest, actionResponse,
+							_language.get(
+								_portal.getHttpServletRequest(actionRequest),
+								"folders-can-not-be-moved-into-articles"));
+
+						return;
+					}
+
+					KBFolder kbFolder = _kbFolderService.getKBFolder(
+						resourcePrimKey);
+
+					if (kbFolder.getParentKBFolderId() ==
+							parentResourcePrimKey) {
+
+						movable = false;
+					}
+				}
 
-			_kbArticleService.moveKBArticle(
-				resourcePrimKey, parentResourceClassNameId,
-				parentResourcePrimKey, priority);
+				if (movable) {
+					_kbFolderService.moveKBFolder(
+						resourcePrimKey, parentResourcePrimKey);
+				}
+			}
```

Comment:
As before, try to remove `movable` and complement the conditions to remove nesting.